Move exir.delegate to PyTorch core to enforce no out-of-tree HOPs

ezyang opened the pull request:
Summary: When HOPs live out of tree, it makes it impossible to make breaking changes to the HOP API. But HOP implementations are deeply entwined with PyTorch internals. Move the HOP into PyTorch tree so that changes are possible.

Test Plan: sandcastle, ossci

Differential Revision: D60674615

Skylion007 commented on torch/_higher_order_ops/executorch_call_delegate.py:
```diff
@@ -0,0 +1,163 @@
+
+
+# pyre-ignore: Missing parameter annotation [2]: Parameter `obj` must have a type other than `Any`.Pyre
+def is_lowered_module(obj: Any) -> bool:
```
this should use a TypeIs or TypeGuard

Skylion007 commented on torch/_higher_order_ops/executorch_call_delegate.py:
```diff
@@ -0,0 +1,163 @@
+# pyre-ignore: Missing parameter annotation [2]: Parameter `obj` must have a type other than `Any`.Pyre
+def is_lowered_module(obj: Any) -> bool:
+    """
+    This function is added to avoid using isinstance(obj, LoweredBackendModule) as it will import LoweredBackendModule, which may cause a circular import.
```
Easier way is to import it only in function

ezyang replied:
I'm happy to apply some of these fixes after landing, I want the import from executorch repo as clean as possible though lol

Skylion007 reviewed (approved):
Please add TypeIs / TypeGuard in a follow up PR.

ezyang committed:
Move exir.delegate to PyTorch core to enforce no out-of-tree HOPs (#132525)

Summary:
X-link: https://github.com/pytorch/executorch/pull/4521


When HOPs live out of tree, it makes it impossible to make breaking changes to the HOP API. But HOP implementations are deeply entwined with PyTorch internals. Move the HOP into PyTorch tree so that changes are possible.

Test Plan: sandcastle, ossci

Differential Revision: D60674615
